src/include/storage/index/bplustree.h

```diff
@@ -0,0 +1,2554 @@
+ */
+template <typename KeyType, typename ValueType, typename KeyComparator = std::less<KeyType>,
+          typename KeyEqualityChecker = std::equal_to<KeyType>,
+          typename ValueEqualityChecker = std::equal_to<ValueType>>
+class BPlusTree : public BPlusTreeBase {
+ public:
+  class BaseNode;
+
+  /** <KeyType, NodeID> pair */
+  using KeyNodePointerPair = std::pair<KeyType, BaseNode *>;
+  /** <KeyType, List of ValueType> pair */
+  using KeyValuePair = std::pair<KeyType, std::list<ValueType> *>;
+  /** <KeyType, ValueType> pair */
+  using KeyElementPair = std::pair<KeyType, ValueType>;
+
+  /**
+   * enum class NodeType - B+ Tree node type
+   */
+  enum class NodeType : int { InnerType = 0, LeafType = 1 };
+
+  ///////////////////////////////////////////////////////////////////
+  // Key Comparison Member Functions
+  ///////////////////////////////////////////////////////////////////
+
+  /**
+   * KeyCmpLess() - Compare two keys for "less than" relation
+   *
+   * If key1 < key2 return true
+   * If not return false
+   */
+  bool KeyCmpLess(const KeyType &key1, const KeyType &key2) const { return key_cmp_obj_(key1, key2); }
+
+  /**
+   * KeyCmpEqual() - Compare a pair of keys for equality
+   *
+   * This functions compares keys for equality relation
+   */
+  bool KeyCmpEqual(const KeyType &key1, const KeyType &key2) const { return key_eq_obj_(key1, key2); }
+
+  /**
+   * KeyCmpGreaterEqual() - Compare a pair of keys for >= relation
+   *
+   * It negates result of keyCmpLess()
+   */
+  bool KeyCmpGreaterEqual(const KeyType &key1, const KeyType &key2) const { return !KeyCmpLess(key1, key2); }
+
+  /**
+   * KeyCmpGreater() - Compare a pair of keys for > relation
+   *
+   * It flips input for keyCmpLess()
+   */
+  bool KeyCmpGreater(const KeyType &key1, const KeyType &key2) const { return KeyCmpLess(key2, key1); }
+
+  /**
+   * KeyCmpLessEqual() - Compare a pair of keys for <= relation
+   */
+  bool KeyCmpLessEqual(const KeyType &key1, const KeyType &key2) const { return !KeyCmpGreater(key1, key2); }
+
+  ///////////////////////////////////////////////////////////////////
+  // Value Comparison Member
+  ///////////////////////////////////////////////////////////////////
+
+  /**
+   * ValueCmpEqual() - Compares whether two values are equal
+   */
+  bool ValueCmpEqual(const ValueType &v1, const ValueType &v2) { return value_eq_obj_(v1, v2); }
+
+  /**
+   * class NodeMetaData - Holds node metadata in an object
+   *
+   * NOTE: We do not count node type as node metadata
+   */
+  class NodeMetaData {
+   public:
+    /** Low and High keys in a leaf or inner node. These are explicitly stored so
+     * that the iterator begin and end can fetch the values easily.
+     */
+    const KeyNodePointerPair *low_key_p_;  /*!< Low Key */
+    const KeyNodePointerPair *high_key_p_; /*!< High Key */
+
+    /** The type of the node */
+    NodeType type_;
+
+    /** This is the height of the node */
+    int depth_;
```

Comment:
Would it be better to comment as "The depth of the node within the tree" instead of using height?

Reply:
Done